### ISSUE TITLE
Gitlab runner feature request

### ISSUE THREAD

**Issue by oplatek** [gitlabhq/terraform-provider-gitlab#49]
One of the obvious features to terraform is Gitlab runner (https://docs.gitlab.com/runner/)

What problem I would it solved?

Currently, I have the CI infrastructure terraformed on AWS and is very easy to register new runner via shell cli and remote-exec but impossible to deregister the runner.

```
 provisioner "remote-exec" {
    connection {
      type = "ssh"
      user = "ec2-user"
      private_key = "${file("PRIVATE_KEY")}"
    }

    inline = [
      "sudo gitlab-runner register -n -u https://gitlab.YOUR_DNS.com -r SECRET)_TOKEN -name your--manually-created-runner-waiting-to-be-terraformed --tag-list ci,test --executor shell",
    ]
  }
```

It has quite a disadvantage that the runners accumulates when I redeploy the infrastructure because I have no means how to delete them.

**Comment by mishak87** [comment]
You can deregister runner by using
```
provisioner "remote-exec" {
  when = "create"
  // your current config
}

provisioner "remote-exec" {
  when = "destroy"
  // ...
  inline = [
    "gitlab-runner unregister --all-runers"
  ]
```

**Comment by lukemgriffith** [comment]
Is implementing this as a resource up for grabs? I'd be keen on getting this in.

**Comment by armsnyder** [comment]
Link to api docs for this: https://docs.gitlab.com/ee/api/runners.html

**Comment by Fuuzetsu** [comment]
I have started work today to include basic support for runners in the provider as it's something we require for our setup.

**Comment by Fuuzetsu** [comment]
The start is here: https://github.com/tsurucapital/terraform-provider-gitlab/tree/gitlab-runner-register

It seems to be able to register/unregister which was mostly what I want. But the rest of the API is not that big so I think I'll do all of it potentially.

**Comment by armsnyder** [comment]
@Fuuzetsu Hey, just checking if you're willing to contribute a PR with what you have, and if you need any help?

**Comment by Fuuzetsu** [comment]
Hi @armsnyder

I did want to include a PR but it's quite low in my list of priorities. I got as far as:

* [updating the upstream GitLab Go SDK](https://github.com/xanzy/go-gitlab/pulls?q=is%3Apr+is%3Aclosed+author%3AFuuzetsu) to expose the appropriate APi
* Code up the terraform provider side to re-provision resource if we change any of the immutable fields but allow changes to stuff API actually lets you to change.

The main thing that is missing is adding acceptance tests as on the first attempt I failed to run these. I have only been able to test it on the version of GitLab we use at work which is some old 12.x version. For example, the `token` field in the runner description has been deprecated on new versions so I'm not confident that my changes will work on GitLab 13.x: I suspect not. I think I may also have hit some strangeness that required to re-run apply but I don't recall anymore what it was and whether I fixed it.

I have published what's on that branch into the terraform registry at https://registry.terraform.io/providers/tsurucapital/gitlab/latest . Probably not advisable that you use it for your production setup.

If someone wants to take what's [on the branch](https://github.com/tsurucapital/terraform-provider-gitlab/tree/gitlab-runner-register), slap some tests on it and submit it, that would be very good. I'd be able to stop using the forked provider.

I don't expect I'll get a chance to do this myself for at least 1-2 months.

**Comment by PatrickRice-KSC** [comment]
I've skimmed past this issue a couple times as I look through old issues, and it feels like while there is a potential new resource here, I do somewhat question the value prop of it. The `gitlab_runner` resource wouldn't be able to return any information that would meaningfully help you set up a GitLab runner on a remote resource as the [Details API](https://docs.gitlab.com/ee/api/runners.html#get-runners-details) doesn't include the authentication token which is necessary in the config.toml file for the runner to actually work. 

The flow I would expect for this kind of a resource would be:

- Register GitLab runner and then get the Authentication Token from the API
- Use the authentication token along with other information to build the config.toml file in terraform
- When standing up new infrastructure (via a separate provider such as azure, aws, or gcp), put that config.toml on the system either via user-data or via cloud storage buckets
- Now your runner is idempotent in that you don't need to re-register and have a bunch of "dead" runner like Fuuzetsu mentioned in the OP.

However, without the authentication token, we fail on the second step of this flow. If we can't register a new runner with the resource, then what would we actually manage within terraform? We can't manage the tags without managing the runner ID, so if the runner ID is created outside of TF (which it seems like it needs to be to create the authentication token), then how does this work?

I'd love for this to work personally, because we re-register runners every time we build infrastructure too and it's an annoying flow, but I'm not sure we can do much here unless the upstream API updates to include the authentication token in the response. So I'd suggest we either close this issue for now (as it's been stale for over a year at this point), or mark it as blocked.

Thoughts?

**Comment by Fuuzetsu** [comment]
@PatrickRice-KSC I don't remember all the details but our setup has evolved into pretty much exactly what you're saying. We have:

* `gitlab_runner_register` resource that gets token from GitLab using the code from https://github.com/tsurucapital/terraform-provider-gitlab/tree/gitlab-runner-register : you only get this token once and then it returns empty string each time after; so just store the token in terraform state and be careful to not override it later; re-applying terraform doesn't change the resource so nothing needs to be reprovisioned
* We generate `config.toml` using the token: what goes here is up to you, we use some custom CloudFormation thing that automatically manages EC2 spot instances for us and run our builds on there, on demand
* the config.toml gets put onto S3: note it only happens if it has changed
* The machine actually running the gitlab-runner software gets informed that file changed, downloads it and reloads

We use some `for_each` etc to have separate runners for separate repos and all this kind of stuff. We also set CI env variables and all that good stuff via terraform.

So I think what I mentioned already holds: someone just needs to take what I had on the branch, polish it, add tests and done.

FWIW we're using GitLab.com now instead of some on-site out of date version and it's all been working fine.

**Comment by timofurrer** [comment]
@PatrickRice-KSC little disclaimer from my side: I need to dig into this when I have some more time than right now.

For now:

> The gitlab_runner resource wouldn't be able to return any information that would meaningfully help you set up a GitLab runner on a remote resource as the [Details API](https://docs.gitlab.com/ee/api/runners.html#get-runners-details) doesn't include the authentication token which is necessary in the config.toml file for the runner to actually work.

Would #638 help here? In any case we don't need the Details API to return a token, as long as we can obtain the token once - it's the case for many other resources, too - like PATs etc ...

**Comment by PatrickRice-KSC** [comment]
Thanks for sending that over @Fuuzetsu  - It looks like some confusion in reading the documentation because the API overloads the `token` term. The token you pass into the API is the Registration Token (which is [documented as such](https://docs.gitlab.com/ee/api/runners.html#register-a-new-runner)) however the token you get in response, _also called `token`_, is the Authorization token (and _not_ documented as such). I was reading the example provided from the API and assuming that `token` in the example was the `token` documented in the parameters. Silly me 😄 ? Looking through your code example is what caused me to go off and test registering a runner that way and using the returned token in a custom config.toml, which sure enough works.

So yes, I can see how this would be used in a downstream flow to build out a config.toml and thus manage runners. I'll take a look through your code - the coding patterns have changed a bit in the main provider since you wrote the code, so I'll see if I can translate and get a new PR submitted to get this into the mainline.

@timofurrer  I think you'd need that to be completed before you could use a resource like this to register a new instance-level runner, but I don't know if it would be required.

**Comment by PatrickRice-KSC** [comment]
OK, I'm going to start in on this PR here shortly.

**Comment by PatrickRice-KSC** [comment]
Im working on a PR for go-gitlab here first -there were some deprecations and such that need to be handled there first (active -> paused, removal of scope). I'll block this for now, but it won't stay blocked long.

**Comment by PatrickRice-KSC** [comment]
Removing block up go-gitlab PR was merged. PR for the tf provider moving again 🙂

**Comment by tspearconquest** [comment]
Hi @PatrickRice-KSC I was about to open an issue requesting a way to manage runner registrations for a different use case, and I saw this issue so I wanted to check with you if our use case will be covered with your PR or should I open a new issue?

We use the Kubernetes executor for our runners, so the entrypoint in the container automatically registers a new runner each time a pod is deleted. This eventually leads to us hitting the maximum runner registrations for our group (around 1000), and we have to manually delete them. So since the entrypoint handles registering them, I understand that the state file won't have the runner's registration token.

In my mind, this could still work because I could define the resource in a way that allows me to run `terraform import` to pull the registration tokens in to the state file, and then I could run `terraform destroy`.

Would your PR be able to cover this use case?

**Comment by PatrickRice-KSC** [comment]
@tspearconquest  - Yeah, I think you could use this resource to fix up your flow. Essentially your entrypoint for the runner manager would need to change from registering the runner to instead inject the authorization token that this PR would expose into your `config.toml` file.

You could do this by injecting the authorization token as an environment variable into your pod definition, by injecting it as a variable in a config map, or the best option would be by storing the authorization token in a secret store somewhere (vault, AWS secrets manager, etc) and having your entrypoint retrieve it from there.

**Comment by PatrickRice-KSC** [comment]
Sorry about the delay on this, PR being submitted today. We had a long blocked based on some breaking changes to the go-gitlab API, then I had a separate PR I wanted to wrap up first :)

**Comment by tspearconquest** [comment]
Awesome thanks! I was looking it over and have a couple of questions.
- Where can I find the runner's ID in the UI?
- I don't see an attribute that could be set as "offline" or "stale" but in the UI, runners can either be online, paused, offline or stale. Can `terraform import` import a runner that is offline or stale? That would be the primary use case for me, is to import our existing runners and delete them when we have too many that show up as offline.

**Comment by PatrickRice-KSC** [comment]
Hey @tspearconquest  - Great questions!

1. In the UI (if you're on the latest version), you can go to CI/CD > Runners, and the list of runners will have the ID as part of the name. If you click on the runner, the ID is present in both the URL and in the title of the runner in the UI. Example:

![image](https://user-images.githubusercontent.com/50960557/166121233-52f3bf12-4e21-4d0c-a8d7-ea07abb7ef27.png)

2. offline and stale will appear in the `status` field. That's a good point, I'll add that field as a computed value (since you can't set the status field). Nice catch! As long as terraform has the ID, it can certainly import them. If you imported them and they weren't present in your config, they would be deleted during your next `apply`.

**Comment by tspearconquest** [comment]
Thank you so much! This will help us immensely! I am definitely looking forward to using this on day 1. :)

**Comment by PatrickRice-KSC** [comment]
I should note - if you import existing runners, terraform will not have the authentication_token in state. This means you won't have the ability to create a working `config.toml` from the resource in terraform ☹️ . This is unfortunately a limitation of GitLab's API - it _only_ return the authentication token from the create API.

**Comment by PatrickRice-KSC** [comment]
Hey all! If you're just watching this issue, we're happy to announce this will be released in the 3.15.0 release! Keep an eye out for that release, and let us know if you have any questions 🙂
